### DISCUSSION TITLE
New to Shopify Functions, should I use rust or JS/TS?

### DISCUSSION THREAD

**Discussion by FriedRiceNoodles** [Shopify/function-examples#518] [Show and tell]
I have experience with JS/TS development, will there be some performance or other gains if I use RUST? How big is the gain if any?

**Comment by davejcameron** [comment]
There are performance gains with using Rust, it's 2-4x faster (as measured by fewer instructions.) Highly recommend investing in Rust if you're planning on releasing a Public app, or expect it to operate on large amounts of data.

**Comment by dnpg** [comment]
@FriedRiceNoodles I agreed with @davejcameron that there are lots of gains using Rust. However,
In my personal experience, I tried starting with Rust, and it has been a pain. The examples in Shopify's docs are soooo simple that it doesn't really reflect real problems we need to solve. I keep getting issues with types and it's hard to find either documentation from Shopify or Rust to try to solve them. I spent more than a week on it to try to get an extension to work with basic functionality.
So after wasting that time, I decided to re-do my work and start with Typescript and I got my code working within minutes.

My plan now is to get all my logic working with typescript and down the track hope to fine the time and better docs to migrate to Rust.

**Reply by FriedRiceNoodles** [reply]
Considering the cost of learning Rust, I think I'll use typescript until I face obvious performance issues.

**Reply by makrisoft** [reply]
hi 👋 @dnpg
I'm very interested in hearing about the real problems you need to solve and the pain points in the way. You mentioned types, did you mean that it is hard to know the structure of the generated code from the macros?

Thanks!
Ernie
